datafusion-jit/src/api.rs

```diff
@@ -0,0 +1,630 @@
+    pub fn create_jit(&self) -> JIT {
+        let symbols = self
+            .state
+            .lock()
+            .extern_funcs
+            .values()
+            .map(|s| (s.name.clone(), s.code))
+            .collect::<Vec<_>>();
+        JIT::new(symbols)
+    }
+}
+
+/// Function builder API. Stores the state while
+/// we are constructing an AST for a function.
+pub struct FunctionBuilder {
+    name: String,
+    params: Vec<(String, JITType)>,
+    ret: Option<(String, JITType)>,
+    fields: VecDeque<HashMap<String, JITType>>,
+    assembler_state: Arc<Mutex<AssemblerState>>,
+}
+
+impl FunctionBuilder {
+    fn new(name: impl Into<String>, assembler_state: Arc<Mutex<AssemblerState>>) -> Self {
+        let mut fields = VecDeque::new();
+        fields.push_back(HashMap::new());
+        Self {
+            name: name.into(),
+            params: Vec::new(),
+            ret: None,
+            fields,
+            assembler_state,
+        }
+    }
+
+    /// Add one more parameter to the function.
+    pub fn param(mut self, name: impl Into<String>, ty: JITType) -> Self {
+        let name = name.into();
+        assert!(!self.fields.back().unwrap().contains_key(&name));
+        self.params.push((name.clone(), ty));
+        self.fields.back_mut().unwrap().insert(name, ty);
+        self
+    }
+
+    /// Set return type for the function. Functions are of `void` type by default if
+    /// you do not set the return type.
+    pub fn ret(mut self, name: impl Into<String>, ty: JITType) -> Self {
+        let name = name.into();
+        assert!(!self.fields.back().unwrap().contains_key(&name));
+        self.ret = Some((name.clone(), ty));
+        self.fields.back_mut().unwrap().insert(name, ty);
+        self
+    }
+
+    /// Enter the function body at start the building.
+    pub fn enter_block(&mut self) -> CodeBlock {
+        self.fields.push_back(HashMap::new());
+        CodeBlock {
+            fields: &mut self.fields,
+            state: &self.assembler_state,
+            stmts: vec![],
+            while_state: None,
+            if_state: None,
+            fn_state: Some(GeneratedFunction {
+                name: self.name.clone(),
+                params: self.params.clone(),
+                body: vec![],
+                ret: self.ret.clone(),
+            }),
+        }
+    }
+}
+
+/// Keep `while` condition expr as we are constructing while loop body.
+struct WhileState {
+    condition: Expr,
+}
+
+/// Keep `if-then-else` state, including condition expr, the already built
+/// then statements (if we are during building the else block).
+struct IfElseState {
+    condition: Expr,
+    then_stmts: Vec<Stmt>,
+    in_then: bool,
+}
+
+impl IfElseState {
+    /// Move the all current statements in the `then` block and move to `else` block.
+    fn enter_else(&mut self, then_stmts: Vec<Stmt>) {
+        self.then_stmts = then_stmts;
+        self.in_then = false;
+    }
+}
+
+/// Code block that
```

Comment:
is there missing doc here? Looks like a suddenly broken sentence.

Reply:
Code block consists of statements and acts as anonymous namespace scope for items and variable declarations.

Thanks for pointed out.